# Problème de répartition équitable de bonus

## Contexte

Un professeur désire calculer les moyennes de ses élèves dans sa matière enseignée. 
Soit le tableau montrant les moyennes sur 20 de ses élèves : 

|N°|Noms|Moyennes|
|:-:|:-:|:-:|
|1|Aya|8|
|2|Souley|15|
|3|Juste|19|
|4|Kouakou|4|
|5|David|9|
|6|Ange|13|

Dans cet établissement, la moyenne de validation d'une matière est égale à 8 sur 20. Malheureusement, Kouakou a obtenu la moyenne de 4, ce qui est insuffisant. Le professeur aimant bien ses élèves, veut tous les faire valider. Alors il décide de faire un bonus de +4 à Kouakou pour le ramener à 8. Mais, les autres élèves sont mécontents, ayant mieux travaillé que Kouakou. La concurrence étant rude dans la classe, ils veulent tous obtenir le même bonus et garder le même écart entre chacun. Cependant, si le professeur applique ce bonus (+4) à Juste, il aura 23. Ce qui est incohérent puisque la moyenne maximale est 20.
La problème qui s'impose est donc la répartition équitable des bonus à tous les élèves de sorte à tous leur permettre de valider cette matière tout en respectant les contraintes de validation à savoir une moyenne comprise entre 8 et 20.

Ce dilemme peut s'appliquer à toutes les situations dans lesquelles l'on veut répartir équitablement des objets à un groupe d'individus tout en respectant certaines limites d'ajout.

Résumé des contraintes :
- La moyenne de validation appartient à [8, 20]
- Les moyennes de ce problème sont comprises dans [4, 19]
- La plage de bonus possibles est donc [1, 4]
- La difficulté : faire en sorte que [4, 19] + [1, 4] -> [8, 20]

## Statistiques de départ

Importation des modules

In [2]:
import numpy as np

Déclaration des variables

In [29]:
liste_notes = [8, 15, 19, 4, 9, 13]
notes = np.array(np.sort(liste_notes), dtype=float)
notes

array([ 4.,  8.,  9., 13., 15., 19.])

In [12]:
VALIDATION_MIN = 8 # note minimale de validation
VALIDATION_MAX = 20 # note maximale de validation
note_min = int(notes.min()) # note minimale de la classe
note_max = int(notes.max()) # note maximale de la classe
bonus_max = VALIDATION_MIN - note_min # bonus maximal possible à ajouter
bonus_min = VALIDATION_MAX - note_max if (note_max + bonus_max > VALIDATION_MAX) else bonus_max # bonus maximal possible à ajouter

Quelques statistiques sur l'échantillon

In [5]:
moyenne_classe = notes.mean() # la moyenne de la classe
moyenne_classe

11.333333333333334

In [6]:
variance_classe = notes.var() # la variance de la classe
variance_classe

24.222222222222225

In [7]:
ecart_type_classe = notes.std() # l'écart-type de la classe
ecart_type_classe

4.921607686744467

## Tentatives de solution

### Solution N°1

In [34]:
liste_temp_1 = notes
liste_temp_1

array([ 4.,  8.,  9., 13., 15., 19.])

In [40]:
liste_bonus_1 = np.linspace(bonus_max, bonus_min, notes.size) # liste des bonus à appliquer
liste_bonus_1

array([4. , 3.4, 2.8, 2.2, 1.6, 1. ])

Application des bonus à chaque élève avec la 1ere méthode

In [41]:
resultat = []
for index, note in enumerate(liste_temp_1):
	if note in notes:
		resultat.append(liste_temp_1[index] + liste_bonus_1[index])
notes_1 = np.array(resultat)
notes_1

array([ 8. , 11.4, 11.8, 15.2, 16.6, 20. ])

Après application des bonus, nous obtenons ainsi le tableau suivant :

|N°|Noms|Moyennes initiales|Bonus|Moyennes finales|
|:-:|:-:|:-:|:-:|:-:|
|1|Juste|19|+1|20|
|2|Souley|15|+1.6|16.6|
|3|Ange|13|+2.2|15.2|
|4|David|9|+2.8|11.8|
|5|Aya|8|+3.4|11.4|
|6|Kouakou|4|+4|8|
||**Moyennes**|**11.33**|**+2.5**|**13.83**|

Quelques statistiques de la solution 1

In [42]:
moyenne_classe_1 = notes_1.mean() # la moyenne de la classe
moyenne_classe_1

13.833333333333334

In [43]:
variance_classe_1 = notes_1.var() # la variance de la classe
variance_classe_1

15.272222222222224

In [44]:
ecart_type_classe_1 = notes_1.std() # l'écart-type de la classe
ecart_type_classe_1

3.9079690661803124

### Solution N°2

In [30]:
liste_temp_2 = np.linspace(note_min, note_max, note_max - note_min + 1)
liste_temp_2

array([ 4.,  5.,  6.,  7.,  8.,  9., 10., 11., 12., 13., 14., 15., 16.,
       17., 18., 19.])

In [31]:
liste_bonus_2 = np.linspace(bonus_max, bonus_min, note_max - note_min + 1)
liste_bonus_2

array([4. , 3.8, 3.6, 3.4, 3.2, 3. , 2.8, 2.6, 2.4, 2.2, 2. , 1.8, 1.6,
       1.4, 1.2, 1. ])

Application des bonus à chaque élève avec la 2e méthode

In [32]:
resultat = []
for index, note in enumerate(liste_temp_2):
	if note in notes:
		resultat.append(liste_temp_2[index] + liste_bonus_2[index])
notes_2 = np.array(resultat)
notes_2

array([ 8. , 11.2, 12. , 15.2, 16.8, 20. ])

Après application des bonus, nous obtenons ainsi le tableau suivant :

|N°|Noms|Moyennes initiales|Bonus|Moyennes finales|
|:-:|:-:|:-:|:-:|:-:|
|1|Juste|19|+1|20|
|2|Souley|15|+1.8|16.8|
|3|Ange|13|+2.2|15.2|
|4|David|9|+3|12|
|5|Aya|8|+3.2|11.2|
|6|Kouakou|4|+4|8|
||**Moyennes**|**11.33**|**+2.53**|**13.86**|

Statistiques de la solution N°2

In [16]:
moyenne_classe_2 = notes_2.mean() # la moyenne de la classe
moyenne_classe_2

13.866666666666667

In [17]:
variance_classe_2 = notes_2.var() # la variance de la classe
variance_classe_2

15.502222222222224

In [18]:
ecart_type_classe_2 = notes_2.std() # l'écart-type de la classe
ecart_type_classe_2

3.937286149395574

## Interprétation des résultats

|Méthodes|Naturelle|Solution 1|Solution 2|
|:-:|:-:|:-:|:-:|
|Note minimale|4|8|8|
|Note maximale|19|20|20|
|Moyenne|11,33|13,83|13,86|
|Variance|24.22|15.27|15.5|
|Ecart-type|4.92|3.91|3.94|

1. La seconde méthode améliore mieux la moyenne des élèves que la première.
2. La première méthode réduit les variances dans l'ensemble. Ce qui est plus avantagieux pour les derniers mais désavantagieux pour les premiers.